### PR TITLE
deleting dangling ports when instance enter error status

### LINKED ISSUE #1404: CAPO leaks ports if machines can not be created (HTTP 500 NoValidHosts)
/kind bug

**What steps did you take and what happened:**

Running CAPO 0.6.3 (plus a no_volume patch but this may be irrelevant to what we see).
If CAPO tries to create a machine and OpenStack returns an error (can't create VM) in this process, capo will not clean up the ports.
This means if one reaches the "breaking point" of what OpenStack can create, CAPO will leak the ports till one reaches the quota, or some other limit.
Ports never get cleaned up on 500 HTTP error msg on VM creation.

The order of operation:
1. md is configured to some value of machines that OpenStack can't create
2. md is stuck in ScalingUp state
3. OpenStack returns HTTP 500 (no valid hosts)
4. CAPO deletes the machine properly from the OpenStack but leaves the port
5. This causes ports to leak and eventually hit the quota

Here are some logs:

In this case 11 replicas is a "breaking point" for Openstack (it is able to create 10 machines but not the 11th).
This is what kicks in the port leak avalanche

```
➜  ~ kubectl get md
NAME                  CLUSTER          REPLICAS   READY   UPDATED   UNAVAILABLE   PHASE       AGE     VERSION
test-md-0    test    11         10      11        1             ScalingUp   11d     v1.21.3
```

This OpenStack was brought to it's knees by CAPI creating excessive ports:
(10 real active VMs, other are down leaked ports):
```
root@2481423024b3:# openstack port list  | grep test-md-0 | grep DOWN | wc -l
457
root@2481423024b3:/# openstack port list  | grep lab-sjc3-blue-md-0 | grep ACTIVE | wc -l
10
```

osm/ma showing:
```
status:
  failureMessage: 'OpenStack instance cannot be created: error creating Openstack
    instance: error creating OpenStack instance 4c18a354-1b05-403f-b228-b6cd8bdb6452,
    status changed to error'
  failureReason: UpdateError
```
OpenStack showing 500 (before capi deletes the port):

```
| fault                               | {'code': 500, 'created': '2022-11-28T18:26:18Z', 'message': 'No valid host was found. There are not enough hosts available.', 'details':
```

### CAPO logs (one example of one failed VM / leftover port):

```
I1128 18:22:15.237468       1 openstackmachine_controller.go:97] controller/openstackmachine "msg"="Machine Controller has not yet set OwnerRef" "name"="test-md-0-rg6wq" "namespace"="tenant-capi-mgmnt" "openStackMachine"="test-md-0-rg6wq" "reconciler group"="infrastructure.cluster.x-k8s.io" "reconciler kind"="OpenStackMachine"
I1128 18:22:15.345805       1 openstackmachine_controller.go:97] controller/openstackmachine "msg"="Machine Controller has not yet set OwnerRef" "name"="test-md-0-rg6wq" "namespace"="tenant-capi-mgmnt" "openStackMachine"="test-md-0-rg6wq" "reconciler group"="infrastructure.cluster.x-k8s.io" "reconciler kind"="OpenStackMachine"
I1128 18:22:15.450520       1 openstackmachine_controller.go:97] controller/openstackmachine "msg"="Machine Controller has not yet set OwnerRef" "name"="test-md-0-rg6wq" "namespace"="tenant-capi-mgmnt" "openStackMachine"="test-md-0-rg6wq" "reconciler group"="infrastructure.cluster.x-k8s.io" "reconciler kind"="OpenStackMachine"
I1128 18:22:15.498447       1 openstackmachine_controller.go:97] controller/openstackmachine "msg"="Machine Controller has not yet set OwnerRef" "name"="test-md-0-rg6wq" "namespace"="tenant-capi-mgmnt" "openStackMachine"="test-md-0-rg6wq" "reconciler group"="infrastructure.cluster.x-k8s.io" "reconciler kind"="OpenStackMachine"
I1128 18:22:16.238332       1 openstackmachine_controller.go:293] controller/openstackmachine "msg"="Bootstrap data secret reference is not yet available" "cluster"="test" "machine"="test-md-0-76d989d445-l7hf7" "name"="test-md-0-rg6wq" "namespace"="tenant-capi-mgmnt" "openStackCluster"="test" "openStackMachine"="test-md-0-rg6wq" "reconciler group"="infrastructure.cluster.x-k8s.io" "reconciler kind"="OpenStackMachine"
I1128 18:22:16.950686       1 openstackmachine_controller.go:293] controller/openstackmachine "msg"="Bootstrap data secret reference is not yet available" "cluster"="test" "machine"="test-md-0-76d989d445-l7hf7" "name"="test-md-0-rg6wq" "namespace"="tenant-capi-mgmnt" "openStackCluster"="test" "openStackMachine"="test-md-0-rg6wq" "reconciler group"="infrastructure.cluster.x-k8s.io" "reconciler kind"="OpenStackMachine"
I1128 18:22:17.674630       1 openstackmachine_controller.go:300] controller/openstackmachine "msg"="Reconciling Machine" "cluster"="test" "machine"="test-md-0-76d989d445-l7hf7" "name"="test-md-0-rg6wq" "namespace"="tenant-capi-mgmnt" "openStackCluster"="test" "openStackMachine"="test-md-0-rg6wq" "reconciler group"="infrastructure.cluster.x-k8s.io" "reconciler kind"="OpenStackMachine"
I1128 18:22:17.674670       1 service.go:70] controller/openstackmachine "msg"="failed to create volume service client, no volume related operations allowed" "cluster"="test" "machine"="test-md-0-76d989d445-l7hf7" "name"="test-md-0-rg6wq" "namespace"="tenant-capi-mgmnt" "openStackCluster"="test" "openStackMachine"="test-md-0-rg6wq" "reconciler group"="infrastructure.cluster.x-k8s.io" "reconciler kind"="OpenStackMachine"
I1128 18:22:18.291380       1 openstackmachine_controller.go:402] controller/openstackmachine "msg"="Machine not exist, Creating Machine" "cluster"="test" "machine"="test-md-0-76d989d445-l7hf7" "name"="test-md-0-rg6wq" "namespace"="tenant-capi-mgmnt" "openStackCluster"="test" "openStackMachine"="test-md-0-rg6wq" "reconciler group"="infrastructure.cluster.x-k8s.io" "reconciler kind"="OpenStackMachine" "Machine"="test-md-0-rg6wq"
I1128 18:22:20.122215       1 recorder.go:103] events "msg"="Normal"  "message"="Created port test-md-0-rg6wq-0 with id 3fa740bc-d456-486e-8341-830ab036a78f" "object"={"kind":"OpenStackMachine","namespace":"tenant-capi-mgmnt","name":"test-md-0-rg6wq","uid":"ea9a23da-bd7f-4363-94b2-0a97a74431d9","apiVersion":"infrastructure.cluster.x-k8s.io/v1alpha5","resourceVersion":"65612058"} "reason"="Successfulcreateport"
E1128 18:22:32.214637       1 openstackmachine_controller.go:493] controller/openstackmachine "msg"="OpenStack instance cannot be created: error creating Openstack instance: error creating OpenStack instance 34528e59-6c1c-4c6b-aec0-eee89fc40db0, status changed to error" "error"="UpdateError" "cluster"="test" "machine"="test-md-0-76d989d445-l7hf7" "name"="test-md-0-rg6wq" "namespace"="tenant-capi-mgmnt" "openStackCluster"="test" "openStackMachine"="test-md-0-rg6wq" "reconciler group"="infrastructure.cluster.x-k8s.io" "reconciler kind"="OpenStackMachine"
I1128 18:22:32.214781       1 recorder.go:103] events "msg"="Warning"  "message"="Failed to create server test-md-0-rg6wq: error creating OpenStack instance 34528e59-6c1c-4c6b-aec0-eee89fc40db0, status changed to error" "object"={"kind":"OpenStackMachine","namespace":"tenant-capi-mgmnt","name":"test-md-0-rg6wq","uid":"ea9a23da-bd7f-4363-94b2-0a97a74431d9","apiVersion":"infrastructure.cluster.x-k8s.io/v1alpha5","resourceVersion":"65612058"} "reason"="Failedcreateserver"
E1128 18:22:32.238193       1 controller.go:317] controller/openstackmachine "msg"="Reconciler error" "error"="error creating Openstack instance: error creating OpenStack instance 34528e59-6c1c-4c6b-aec0-eee89fc40db0, status changed to error" "name"="test-md-0-rg6wq" "namespace"="tenant-capi-mgmnt" "reconciler group"="infrastructure.cluster.x-k8s.io" "reconciler kind"="OpenStackMachine"
I1128 18:22:32.924530       1 openstackmachine_controller.go:275] controller/openstackmachine "msg"="Not reconciling machine in failed state. See openStackMachine.status.failureReason, openStackMachine.status.failureMessage, or previously logged error for details" "cluster"="test" "machine"="test-md-0-76d989d445-l7hf7" "name"="test-md-0-rg6wq" "namespace"="tenant-capi-mgmnt" "openStackCluster"="test" "openStackMachine"="test-md-0-rg6wq" "reconciler group"="infrastructure.cluster.x-k8s.io" "reconciler kind"="OpenStackMachine"
I1128 18:22:33.638860       1 openstackmachine_controller.go:202] controller/openstackmachine "msg"="Reconciling Machine delete" "cluster"="test" "machine"="test-md-0-76d989d445-l7hf7" "name"="test-md-0-rg6wq" "namespace"="tenant-capi-mgmnt" "openStackCluster"="test" "openStackMachine"="test-md-0-rg6wq" "reconciler group"="infrastructure.cluster.x-k8s.io" "reconciler kind"="OpenStackMachine"
I1128 18:22:33.638902       1 service.go:70] controller/openstackmachine "msg"="failed to create volume service client, no volume related operations allowed" "cluster"="test" "machine"="test-md-0-76d989d445-l7hf7" "name"="test-md-0-rg6wq" "namespace"="tenant-capi-mgmnt" "openStackCluster"="test" "openStackMachine"="test-md-0-rg6wq" "reconciler group"="infrastructure.cluster.x-k8s.io" "reconciler kind"="OpenStackMachine"
I1128 18:22:36.153249       1 openstackmachine_controller.go:265] controller/openstackmachine "msg"="Reconciled Machine delete successfully" "cluster"="test" "machine"="test-md-0-76d989d445-l7hf7" "name"="test-md-0-rg6wq" "namespace"="tenant-capi-mgmnt" "openStackCluster"="test" "openStackMachine"="test-md-0-rg6wq" "reconciler group"="infrastructure.cluster.x-k8s.io" "reconciler kind"="OpenStackMachine"
I1128 18:22:36.153332       1 recorder.go:103] events "msg"="Normal"  "message"="Deleted server test-md-0-rg6wq with id 34528e59-6c1c-4c6b-aec0-eee89fc40db0" "object"={"kind":"OpenStackMachine","namespace":"tenant-capi-mgmnt","name":"test-md-0-rg6wq","uid":"ea9a23da-bd7f-4363-94b2-0a97a74431d9","apiVersion":"infrastructure.cluster.x-k8s.io/v1alpha5","resourceVersion":"65612239"} "reason"="Successfuldeleteserver"
E1128 18:22:36.170527       1 controller.go:317] controller/openstackmachine "msg"="Reconciler error" "error"="openstackmachines.infrastructure.cluster.x-k8s.io \"test-md-0-rg6wq\" not found" "name"="test-md-0-rg6wq" "namespace"="tenant-capi-mgmnt" "reconciler group"="infrastructure.cluster.x-k8s.io" "reconciler kind"="OpenStackMachine"
```

This is not the correct behavior and CAPO should clean up the ports if the ma/osm is in error state.
Cleanup of the VM works, but it seems we would delete  ports only on cluster deletion (or some other err scenarios), but not this one.

**What did you expect to happen:**

In the case of 500 message (or any failure to create VM), corresponding port should be deleted.
It may be that CAPO is looking at the server properties in the vm creation flow, but in the case of 500 err message, there are  not port data to be fetched (looking at the code for this).

**Anything else you would like to add:**

### According to instance.go we do garbage collect the ports.
I would have expected that this would work, as the logs do indicate that we detect `infrav1.InstanceStateError` as the instance state, so I would expect that the garbage collection kicks in here:

https://github.com/anaperic/cluster-api-provider-openstack/blob/0648aa47c0396198cf55aaf23de6e7b39bfa5d82/pkg/cloud/services/compute/instance.go#L580-L581

### One similar issues is fixed but only on cluster delete it seems (not tried)

I did see this issue and a fix, that we did not try yet, but it seems this solves only the cluster delete port deletion.
And here we are talking about port deletion upon unsuccessful VM creation.

https://github.com/kubernetes-sigs/cluster-api-provider-openstack/issues/1254
There was a previous bug for port leak but this is merged back in 2021 and the 0.6.3 should have this fix:

https://github.com/kubernetes-sigs/cluster-api-provider-openstack/pull/798/files#diff-5e7dbfd52305a21a11fb94fab201062ae2c7b851db93c2a8443a1d82d7aa7936

**Environment:**

- Cluster API Provider OpenStack version (Or `git rev-parse HEAD` if manually built):
(https://github.com/kubernetes-sigs/cluster-api-provider-openstack/issues/1345)

Manually patched (no_volume PR patch) on top of:
```
commit e103e434b851a2a017445cd581de2145a7ee09e9 (origin/release-0.6, release-0.6)
Merge: ecc4cb10 3f35be7a
Author: ji chen <jichenjc@cn.ibm.com>
Date:   Mon Sep 5 15:45:59 2022 +0800

    Merge pull request #1336 from apricote/remove-myself-0.6

    🌱 remove apricote from reviewers
```

- Cluster-API version: 1.2.1
- OpenStack version: xena
- Minikube/KIND version: n/a
- Kubernetes version (use `kubectl version`): 1.21.3
- OS (e.g. from `/etc/os-release`): Ubuntu 20.04.5 LTS

### PR DESCRIPTION
**Pull request by jichenjc**
**What this PR does / why we need it**:

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #1404 

**Special notes for your reviewer**:

1. Please confirm that if this PR changes any image versions, then that's the sole change this PR makes.

**TODOs**:

- [ ] squashed commits
- if necessary:
  - [ ] includes documentation
  - [ ] adds unit tests

/hold

### CODE REVIEW

**Comment by mdbooth** [comment]
It should be possible to test this with #1391

**Comment by jichenjc** [comment]
ok, I think we have this code logic (but only test with UT), so this is to switch ERROR VM delete port from let VM delete back to delete port directly in code logic..

File: pkg/cloud/services/compute/instance.go [hunk 1]
```diff
@@ -280,6 +280,10 @@ func (s *Service) createInstanceImpl(eventObject runtime.Object, openStackCluste
 	})
 	if err != nil {
 		record.Warnf(eventObject, "FailedCreateServer", "Failed to create server %s: %v", createdInstance.Name(), err)
+
+		// we failed to create instance, now with this server = nil, we will let the defer function handle the
+		// defer action to clean up all the ports as it's in infrav1.InstanceStateError status
```
Comment (seanschneeweiss) [inline pkg/cloud/services/compute/instance.go:285]: How can we make sure that the VM is in InstaneStateError and we do not delete the port of a working VM? There might be other errors, such as overloaded API or connection time out.

We observed some cases where the port of the existing and ready VM was deleted after the API was not reachable, responded with error 500 or a connection timeout / connection refused. I haven't found the exact problem in the code yet - but I think it is somehow related to the port deletion within the `defer`.

Comment (jichenjc) [reply pkg/cloud/services/compute/instance.go:285]: >>>such as overloaded API or connection time out.
>>>responded with error 500 or a connection timeout / connection refused.

API return 500 indicate API layer reported `internal error` that didn't handle the exception well
but the VM enter ERROR status due to something wrong ,so not combined condition if I understand correctly

the logic here should only handle the `revert` action of create instance, which is , something went wrong and we don't want to keep those resources to be orphan resource ..

Comment (mdbooth) [reply pkg/cloud/services/compute/instance.go:285]: This isn't the correct place to do this, same as the existing port cleanup, tbh. This really needs to be done in DeleteInstance. Anything else means that any interruption in the controller, for example a restart or some unrelated error, can result in resource leakage.

**Pull request closed without merging**